### ISSUE TITLE
Extend "discriminate" so it can handle impossible equations such as n = S n

### ISSUE THREAD

**Issue by jfehrle** [rocq-prover/rocq#19504]
As @herbelin  said [here](https://github.com/coq/coq/pull/19475#pullrequestreview-2278374479):
> On the other side, I'm unsure we have to be proud that cyclic equalities are not automatically found contradictory. If not by `discriminate`, there should be a tactic to solve them automatically (can't Equations solve them?).

See example of the issue in #19475

IMO much simpler for users if we can improve `discriminate` rather than introduce a new tactic.

**Comment by herbelin** [comment]
Does such code already exist in Equations? I found instances of it and when trying an example, it seemed to confirm that we need to prove `NoCycle` manually:
```coq
From Equations Require Import Equations.      
Inductive T : list nat -> list nat -> Type := C l : T l (cons 1 l).
Equations f l (a:T l l) : False := .
> ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Error: [noConfusion] Cannot find an instance of NoCycle for type (list nat)
```
 I found nothing of that sort in Elpi too.

According to C. McBride, H. Goguen, and J. McKinna, "A Few Constructions on Constructors", that should be doable using `Below` which seems to be defined in Equations. @mattam82, help!

**Comment by herbelin** [comment]
After some experiments, I realize that it would be very easy to extend injection/discriminate to prove acyclicity. Assume the equality is `x = phi(y,x)` with `phi` a nesting of constructors. Then, the proof of `forall y x, x <> phi(y,x)` is:
```coq
fix f y x {struct x} : x <> phi(y,x) :=
  fun H =>
  match H in _ = l return match l with
                          | phy(y,x) => x <> phi(y,x)
                          | _ => True
                          end
  with
  | eq_refl =>
    match l return match l with
                   | phi(y,x) => x <> phi(y,x)
                   | _ => True
                   end
    with
    | phi(y,x) => f y x
    | _ => I
    end
  end H).
```
that is, using P. Boutillier's notations:
```coq
fix f y x {struct x} : x <> phi(y,x) :=
  fun H =>
  match H in _ = phy(y,x) return x <> phi(y,x) with
  | eq_refl =>
    match l with
    | phi(y,x) => f y x
    | _ => I
    end
  end H).
```
For instance, the proof of `l <> a :: b :: l` is:
```coq
Open Scope list_scope.
Fixpoint acyclic {A} (a b:A) l : l <> a :: b :: l :=
  fun H =>
  match H in _ = a :: b :: l => l <> a :: b :: l with
  | eq_refl =>
    match l with
    | a :: b :: l => acyclic a b l
    | _ => I
    end
  end H).
```

**Comment by silene** [comment]
Note that, by naming the return clause (as often suggested in papers by Monin et al), you get an equivalent but much more readable proof term (and possibly easier to typecheck). On your example, the proof would become as follows:
```coq
Definition acyclic T G :=
  let diag y := match y with a :: b :: l => l = a :: b :: l -> G | _ => True end in
  fix aux (a b:T) l (H: l = a :: b :: l): G :=
  match H in _ = y return diag y with
  | eq_refl =>
    match l with
    | a :: b :: l => aux a b l
    | _ => I
    end
  end H.
```

**Comment by herbelin** [comment]
Adding "good first issue" tag in the following sense:
- does not probably require more than 40 lines (around `Equality.find_positions)
- requires however to understand how strong elimination of `match` works

**Comment by mattam82** [comment]
I don't believe it's that simple to generate a fix etc in general. Equations derives it from a NoCycle class indeed. NoCycle is implied by Wellfoundedness of the subterm relation, for example. I.e. to show that an equality is false one shows that `t` cannot be a subterm of `C t` essentially.

**Comment by herbelin** [comment]
My impression is precisely that what Conor was doing with an explicit subterm relation comes for free when relying on the guard.

**Comment by JasonGross** [comment]
I expect you need a more complicated strategy when the subterm is under binders, for example if we have `C : (A -> T) -> T`.  But it seems that when the subterm is not under binders, this technique should work fine.
